### DISCUSSION TITLE
请问在编译时将SQLite3打开后，如何使用它的消息持久化？是否需要自行配置？

### DISCUSSION THREAD

**Discussion by HelliWrold1** [nanomq/nanomq#606] [Q&A] [answered]
编译完成并启动后，使用MQTTX连接nanomq，报错```segmentation fault```
![屏幕截图 2022-05-20 151156](https://user-images.githubusercontent.com/37153307/169474308-86c2ce94-882d-4ce5-a8d6-97cbc42860fc.jpg)

**Comment by JaylinYu** [comment]
hi Sqlite3现在是作为QoS消息缓存功能（未发送成功收到ack的消息会缓存，而不是全量消息持久化），不需要另外配置。

看你的情况像是启动了2个nanomq  第二个被阻止了，请提供core dump文件

**Comment by HelliWrold1** [comment]
好的，谢谢。我的确是启动了两个，但是我在启动自己编译的nanomq时，将使用发行包安装的nanomq停止运行了。另外core dump文件，我一直没有打开生成，刚刚打开后，启动自己编译的nanomq，再次连接，然后自行断开，提示segmentation fault，core dump文件只有“core”
另外，我想要使用nanomq将数据本地化存储，然后进行分析和处理，最后将需要上传的消息发给云端服务器，所以目前版本是不支持所有数据或用户指定数据的本地存储的是吗？
![屏幕截图 2022-05-20 173405](https://user-images.githubusercontent.com/37153307/169500071-9878270b-f5a1-4ecc-9203-4c324ddd6ba9.jpg)

**Comment by JaylinYu** [comment]
这个段错误 core dump确认为只有用MQTTX才会触发的bug  今天会修复

**Comment by JaylinYu** [comment]
全量数据的本地存储，nanomq支持使用fdb，如果您想要使用sqlite进行全量存储，可以写一个Feature Request。或者去discussion投票一下https://github.com/emqx/nanomq/discussions/540

**Comment by HelliWrold1** [comment]
> 全量数据的本地存储，nanomq支持使用fdb，如果您想要使用sqlite进行全量存储，可以写一个Feature Request。或者去discussion投票一下#540

是foundationdb吗？我看到14天前您将其添加为新特性，目前的公开版本已经支持foundationdb了吗？

**Comment by JaylinYu** [comment, marked as answer]
> > 全量数据的本地存储，nanomq支持使用fdb，如果您想要使用sqlite进行全量存储，可以写一个Feature Request。或者去discussion投票一下#540
> 
> 是foundationdb吗？我看到14天前您将其添加为新特性，目前的公开版本已经支持foundationdb了吗？

是的 其实现在最新的master分支已经支持了，但要有完整文档和教程还请等待6月初的0.8.0

**Reply by Youssion** [reply]
hi 文档“配置说明(V014)”中已有sqlite3的规则引擎配置项，是否指新版本可以使用sqlite3进行数据持久化？

**Reply by JaylinYu** [reply]
yes 其实还支持mysql，但一般人我不告诉他

**Reply by Youssion** [reply]
sry 在使用0.16.5版本时，我将nanomq.conf中rules一段更改如下并向‘’testtopic发送了符合规则的数据，但未能成功将数据存入sqlite3，请问是什么问题呢？
![图片](https://user-images.githubusercontent.com/75729225/230395220-9e995395-e619-41ac-9983-38fb5d367243.png)

**Reply by xinyi-xs** [reply]
我在本地测试了图中的规则，是可以正常存储，但是这里可能存在两个问题吗，一个是你的 path 的路径是否合法？，另外一个是你发布的 payload 的数据是不是符合规则，这个如果方便的话，可以贴出来我看下。

**Reply by Youssion** [reply]
![图片](https://user-images.githubusercontent.com/75729225/230537817-f782c145-6755-46f1-ac24-0ca865e5a3eb.png)

![图片](https://user-images.githubusercontent.com/75729225/230538034-93925ad8-9bb9-4d9d-b76a-81c889e91eb7.png)

以上我在docker中用sqlite3创建了'test.db'，启动nanomq能成功连接和订阅，向'abc'发送JSON数据试图命中"select topic, payload from \"abc\" "，但查看sqlite发现数据并未存入。
（docker中时间有误，实际是刚截的图）

**Reply by Youssion** [reply]
hi 请问是否是数据库设置有误

**Reply by xinyi-xs** [reply]
这里我们没有对数据进行过什么设置的，看你的消息应该也是符合规则的，清空下 db 文件里的 table 重新试下呢？

**Comment by JaylinYu** [comment]
我先close这个issue了 转成数据持久化的讨论
